zcash_primitives/src/zip32.rs

```diff
@@ -159,6 +160,18 @@ impl DiversifierKey {
         Self::try_diversifier_internal(&ff, j)
     }
 
+    /// Decrypts a diversifier using this diversifier key to obtain the
+    /// diversifier index from which it was originally created.
+    pub fn diversifier_index(&self, d: &Diversifier) -> Option<DiversifierIndex> {
```

Comment:
And since we no longer use `?`, we can always return the diversifier index (unlike when deriving the diversifier, we know here that `Diversifier` is valid).
```suggestion
    pub fn diversifier_index(&self, d: &Diversifier) -> DiversifierIndex {
```